### PR TITLE
DT-1376: ✨ Add in roles and group search and limit group manager results

### PR DESCRIPTION
**Pull request by petergphillips**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/auth/model/UserFilter.kt [hunk 1]
```diff
@@ -12,13 +12,11 @@ import javax.persistence.criteria.Root
...
 class UserFilter(
   name: String? = null,
-  roleCode: String? = null,
-  groupCode: String? = null,
+  val roleCodes: List<String>? = null,
+  val groupCodes: List<String>? = null,
 ) : Specification<User> {
 
   private val name: String? = if (name.isNullOrBlank()) null else name.trim()
-  private val roleCode: String? = if (roleCode.isNullOrBlank()) null else roleCode.trim()
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/auth/model/UserFilter.kt:20]: `trim` moved lower down to `forEach`

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/auth/model/UserFilter.kt [hunk 2]
```diff
@@ -44,7 +53,7 @@ class UserFilter(
     val pattern = "%" + name!!.replace(',', ' ').replace(" [ ]*".toRegex(), "% %") + "%"
     orBuilder.add(cb.like(root.get("email"), format(pattern)))
     orBuilder.add(cb.like(root.get("username"), pattern.toUpperCase()))
-    val personJoin = root.join<Any, Any>("person", JoinType.LEFT)
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/auth/model/UserFilter.kt:47]: no idea why this was `LEFT` when `person` details are mandatory for auth users

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/maintain/AuthUserService.kt [hunk 3]
```diff
@@ -91,15 +91,32 @@ class AuthUserService(
...
+    roleCodes: List<String>?,
+    groupCodes: List<String>?,
+    pageable: Pageable,
+    searcher: String,
+    authorities: Collection<GrantedAuthority>,
+  ): Page<User> {
+    val groupSearchCodes = if (authorities.any { it.authority == "ROLE_MAINTAIN_OAUTH_USERS" }) {
+      groupCodes
+    } else {
+      val assignableGroupCodes = authUserGroupService.getAssignableGroups(searcher, authorities).map { it.groupCode }
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/maintain/AuthUserService.kt:105]: limit group managers to being able to only search in their group

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/resource/api/AuthUserController.kt [hunk 4]
```diff
@@ -134,16 +134,19 @@ class AuthUserController(
       value = "Sort column and direction, eg sort=lastName,desc"
     )
   )
+  @PreAuthorize("hasAnyRole('ROLE_MAINTAIN_OAUTH_USERS', 'ROLE_AUTH_GROUP_MANAGER')")
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/resource/api/AuthUserController.kt:137]: limit search to group managers and auth admins

File: src/test/kotlin/uk/gov/justice/digital/hmpps/oauth2server/auth/repository/UserRepositoryTest.kt [hunk 5]
```diff
@@ -99,7 +99,7 @@ class UserRepositoryTest {
     val retrievedEntity = repository.findByUsername(transientEntity.username).orElseThrow()
 
     // equals only compares the business key columns
-    assertThat(retrievedEntity).isEqualToIgnoringGivenFields(transientEntity, "authorities")
+    assertThat(retrievedEntity).usingRecursiveComparison().ignoringFields("authorities").isEqualTo(transientEntity)
```
Comment (petergphillips) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/oauth2server/auth/repository/UserRepositoryTest.kt:102]: remove deprecated api usage warning

File: src/test/kotlin/uk/gov/justice/digital/hmpps/oauth2server/auth/repository/UserRepositoryTest.kt [hunk 6]
```diff
@@ -317,11 +331,11 @@ class UserRepositoryTest {
       .containsExactly(
         "AUTH_NO_EMAIL",
         "AUTH_MFA_NOEMAIL_USER",
-        "NOMIS_ENABLED_AUTH_DISABLED",
+        "AUTH_MFA_NOTEXT_USER",
```
Comment (petergphillips) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/oauth2server/auth/repository/UserRepositoryTest.kt:334]: ordering changes due to order by now on search results

File: src/test/resources/uk/gov/justice/digital/hmpps/oauth2server/resource/api/auth_user_search_group_manager.json [hunk 7]
```diff
@@ -0,0 +1,39 @@
+{
+  "content": [
```
Comment (petergphillips) [inline src/test/resources/uk/gov/justice/digital/hmpps/oauth2server/resource/api/auth_user_search_group_manager.json:2]: only one result as group manager can only see one of the users

**Review by Mjwillis — APPROVED** [review]
lgtm

**Pull request merged**